cmd/createtree/main_test.go

```diff
@@ -43,147 +61,113 @@ func marshalAny(p proto.Message) *any.Any {
-	emptyAddr.addr = ""
-
-	invalidEnumOpts := *validOpts
-	invalidEnumOpts.treeType = "LLAMA!"
-
-	invalidKeyTypeOpts := *validOpts
-	invalidKeyTypeOpts.privateKeyType = "LLAMA!!"
-
-	emptyPEMPath := *validOpts
-	emptyPEMPath.pemKeyPath = ""
-
-	emptyPEMPass := *validOpts
-	emptyPEMPass.pemKeyPass = ""
-
-	pemKeyOpts := *validOpts
-	pemKeyOpts.privateKeyType = "PEMKeyFile"
-	pemKeyTree := *defaultTree
-	pemKeyTree.PrivateKey, err = ptypes.MarshalAny(&keyspb.PEMKeyFile{
-		Path:     pemPath,
-		Password: pemPassword,
+	runTest(t, []*testCase{
+		{
+			desc: "validOpts",
+			// runTest sets mandatory options, so no need to provide a setFlags func.
+			wantTree: defaultTree,
+		},
+		{
+			desc: "nonDefaultOpts",
+			setFlags: func() {
+				*treeType = nonDefaultTree.TreeType.String()
+				*signatureAlgorithm = nonDefaultTree.SignatureAlgorithm.String()
+				*displayName = nonDefaultTree.DisplayName
+				*description = nonDefaultTree.Description
+			},
+			wantTree: &nonDefaultTree,
+		},
+		{
+			desc: "mandatoryOptsNotSet",
+			// Undo the flags set by runTest, so that mandatory options are no longer set.
+			setFlags: resetFlags,
+			wantErr:  true,
+		},
+		{
+			desc:     "emptyAddr",
+			setFlags: func() { *adminServerAddr = "" },
+			wantErr:  true,
+		},
+		{
+			desc:     "invalidEnumOpts",
+			setFlags: func() { *treeType = "LLAMA!" },
+			wantErr:  true,
+		},
+		{
+			desc:     "invalidKeyTypeOpts",
+			setFlags: func() { *privateKeyFormat = "LLAMA!!" },
+			wantErr:  true,
+		},
+		{
+			desc:      "createErr",
+			createErr: errors.New("create tree failed"),
+			wantErr:   true,
+		},
 	})
-	if err != nil {
-		t.Fatalf("MarshalAny(PEMKeyFile): %v", err)
-	}
+}
 
-	pkcs11Opts := *validOpts
-	pkcs11Opts.privateKeyType = "PKCS11ConfigFile"
-	pkcs11Opts.pkcs11ConfigPath = "testdata/pkcs11-conf.json"
-	pkcs11Tree := *defaultTree
-	pkcs11Tree.PrivateKey, err = ptypes.MarshalAny(&keyspb.PKCS11Config{
-		TokenLabel: "log",
-		Pin:        "1234",
-		PublicKey: `-----BEGIN PUBLIC KEY-----
-MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC7/tWwqUXZJaNfnpvnqiaeNMkn
-hKusCsyAidrHxvuL+t54XFCHJwsB3wIlQZ4mMwb8mC/KRYhCqECBEoCAf/b0m3j/
-ASuEPLyYOrz/aEs3wP02IZQLGmihmjMk7T/ouNCuX7y1fTjX3GeVQ06U/EePwZFC
-xToc6NWBri0N3VVsswIDAQAB
------END PUBLIC KEY-----
-`,
-	})
-	if err != nil {
-		t.Fatalf("MarshalAny(PKCS11Config): %v", err)
+// runTest executes the createtree command against a fake Trillian server for
+// each of the provided test cases, and checks that the tree in the request is
+// as expected, or an expected error occurs.
+// It changes a couple of flags from their default values before performing the
+// test. It sets the adminServerAddr flag to point to the fake server and sets
+// the privateKeyFormat flag to "test". It also registers a "test" key handler.
+// Finally, it calls the test's setFlags func (if provided) to allow it to
+// change flags specific to the test, then performs the test.
+func runTest(t *testing.T, tests []*testCase) {
```

Comment:
a slightly more descriptive name would be great - and the test should probably only run one test case at a time